src/olympia/blocklist/utils.py

```diff
@@ -4,68 +4,92 @@
 from olympia import amo
 from olympia.activity import log_create
 from olympia.users.utils import get_task_user
+from olympia.versions.models import Version
 
 
 log = olympia.core.logger.getLogger('z.amo.blocklist')
 
 
-def add_version_log_for_blocked_versions(obj, al, submission_obj=None):
-    from olympia.activity.models import VersionLog
-
-    VersionLog.objects.bulk_create(
-        [
-            VersionLog(activity_log=al, version=version)
-            for version in obj.addon_versions
-            if version.is_blocked
-            or (submission_obj and version.id in submission_obj.changed_version_ids)
-        ]
-    )
-
-
 def block_activity_log_save(
     obj,
     change,
     submission_obj=None,
 ):
     action = amo.LOG.BLOCKLIST_BLOCK_EDITED if change else amo.LOG.BLOCKLIST_BLOCK_ADDED
-    version_ids = sorted(ver.id for ver in obj.addon_versions if ver.is_blocked)
+    addon_versions = {ver.id: ver.version for ver in obj.addon_versions}
+    blocked_versions = sorted(
+        ver.version for ver in obj.addon_versions if ver.is_blocked
+    )
+    changed_version_ids = (
+        [v_id for v_id in submission_obj.changed_version_ids if v_id in addon_versions]
+        if submission_obj
+        else sorted(ver.id for ver in obj.addon_versions if ver.is_blocked)
+    )
+    changed_versions = sorted(addon_versions[ver_id] for ver_id in changed_version_ids)
+
     details = {
         'guid': obj.guid,
-        'versions': version_ids,
+        'blocked_versions': blocked_versions,
+        'added_versions': changed_versions,
         'url': obj.url,
         'reason': obj.reason,
-        'comments': f'{len(version_ids)} versions blocked.',
+        'comments': f'{len(changed_versions)} versions added to block; '
+        f'{len(blocked_versions)} total versions now blocked.',
     }
     if submission_obj:
         details['signoff_state'] = submission_obj.SIGNOFF_STATES.get(
             submission_obj.signoff_state
         )
         if submission_obj.signoff_by:
             details['signoff_by'] = submission_obj.signoff_by.id
-    addon = obj.addon
-    al = log_create(action, addon, obj.guid, obj, details=details, user=obj.updated_by)
+
+    log_create(action, obj.addon, obj.guid, obj, details=details, user=obj.updated_by)
+    for version_id in changed_version_ids:
```

Comment:
I wonder if it's worth trying to create a single `ActivityLog` with multiple versions arguments like we do in reviewer tools for actions dealing with multiple versions. That still results in multiple `VersionLog` (which may be an issue when it comes to performance - the `bulk_create()` that was previously there was probably helping, by making a single query for all).

Maybe a follow-up.

Reply:
I looked at the `reject_multiple_versions` action - it seemed the most analogous - and it just loops through and called `log_create` individually.

Reply:
It loops over `user_and_versions` (and `action_to_record`) because it processes delayed rejections into actual ones, and there might be a mix of content/code rejections made by several reviewers. But in the general case, it only records a single `ActivityLog` (it passes `versions` to `self.log_action()` in the review helper). Inside `ActivityLog.create()` it then creates as many `VersionLog` as necessary (*), but only a single `ActivityLog` for multiple versions.

(*) We could `bulk_create()` them too ...

Reply:
https://github.com/mozilla/addons-server/issues/20948